src/rustzcash.rs

```diff
@@ -191,6 +219,103 @@ pub extern "system" fn librustzcash_merkle_hash(
     write_le(tmp, &mut result[..]);
 }
 
+#[no_mangle] // ToScalar
+pub extern "system" fn librustzcash_to_scalar(
+    input: *const [c_uchar; 64],
+    result: *mut [c_uchar; 32],
+) {
+    // Should be okay, because caller is responsible for ensuring
+    // the pointer is a valid pointer to 32 bytes, and that is the
+    // size of the representation
+    let mut scalar = <Bls12 as JubjubEngine>::Fs::to_uniform(unsafe { &(&*input)[..] }).into_repr();
+
+    let result = unsafe { &mut *result };
+
+    scalar.as_mut().reverse();
+    for b in scalar.as_mut() {
+        *b = swap_bits_u64(*b);
+    }
+
+    scalar
+        .write_be(&mut result[..])
+        .expect("length is 32 bytes");
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_ask_to_ak(
+    ask: *const [c_uchar; 32],
+    result: *mut [c_uchar; 32],
+) {
+    let ask = unsafe { &*ask };
+    let ak = fixed_scalar_mult(ask, FixedGenerators::SpendingKeyGenerator);
+
+    let result = unsafe { &mut *result };
+
+    ak.write(&mut result[..]).expect("length is 32 bytes");
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_nsk_to_nk(
+    nsk: *const [c_uchar; 32],
+    result: *mut [c_uchar; 32],
+) {
+    let nsk = unsafe { &*nsk };
+    let nk = fixed_scalar_mult(nsk, FixedGenerators::ProofGenerationKey);
+
+    let result = unsafe { &mut *result };
+
+    nk.write(&mut result[..]).expect("length is 32 bytes");
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_crh_ivk(
```

Comment:
Psst! There's an API for this in sapling-crypto.

https://github.com/zcash-hackworks/sapling-crypto/blob/master/src/primitives/mod.rs#L89

Reply:
I can see why you might not want to use it though. It looks like you do the right thing here, so meh, it's okay.

Reply:
Well, kinda. I think you should do what my code does, interpret it as a scalar (it should always work) and then use a `write_fs` afterwards, which should be implemented to write out in little endian bit order.

Reply:
Using that API requires six additional conversion read/write operations (to convert ak and nk to scalars and back, and to convert ivk to a Point and back). It didn't seem worth the cycles.

Reply:
Once this has been rebased, we should test this function against the test vectors.